File: ios/ReservationApp/ReservationApp/Presentation/ViewModels/MakeReservationViewModel.swift
//
//  MakeReservationViewModel.swift
//  ReservationApp
//
//  Created by Mehmet Ali Kısacık on 13.11.2025.
//

import Foundation
import SwiftUI
import Combine

// Date selection model
struct SelectedDate: Identifiable, Equatable {
    let id = UUID()
    let date: String // YYYY-MM-DD format
    var mealTimeSlotId: Int?
    
    static func == (lhs: SelectedDate, rhs: SelectedDate) -> Bool {
        lhs.date == rhs.date
    }
}

@MainActor
class MakeReservationViewModel: ObservableObject {
    @Published var currentStep: Int = 1
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var successMessage: String?
    
    // Reservation data
    @Published var selectedDates: [SelectedDate] = []
    @Published var selectedRestaurant: Restaurant?
    @Published var selectedMenuType: MenuType?
    @Published var selectedMenu: Menu?
    @Published var selectedMeal: Meal?
    @Published var appetizer: Bool = false
    
    // Available options
    @Published var restaurants: [Restaurant] = []
    @Published var mealTimeSlots: [MealTimeSlot] = []
    @Published var availableMenus: [Menu] = []
    
    private let homeRepository: HomeRepositoryProtocol
    private let reservationRepository: ReservationRepositoryProtocol
    
    let maxDateSelections = 2
    
    init(
        homeRepository: HomeRepositoryProtocol = HomeRepository(),
        reservationRepository: ReservationRepositoryProtocol = ReservationRepository()
    ) {
        self.homeRepository = homeRepository
        self.reservationRepository = reservationRepository
    }
    
    func loadInitialData() async {
        isLoading = true
        errorMessage = nil
        
        async let restaurantsTask = homeRepository.getRestaurants()
        async let mealTimeSlotsTask = homeRepository.getMealTimeSlots()
        
        do {
            let (restaurantsResult, mealTimeSlotsResult) = try await (restaurantsTask, mealTimeSlotsTask)
            self.restaurants = restaurantsResult
            self.mealTimeSlots = mealTimeSlotsResult
        } catch {
            errorMessage = "Veriler yüklenirken bir hata oluştu."
        }
        
        isLoading = false
    }
    
    func loadMenus() async {
        guard let firstDate = selectedDates.first?.date,
              let restaurant = selectedRestaurant,
              let menuType = selectedMenuType else {
            return
        }
        
        isLoading = true
        errorMessage = nil
        
        do {
            // Fetch menus for the first date
            // Note: Backend expects date, restaurantId, and menuType
            let menus = try await homeRepository.getMenus(
                date: firstDate,
                restaurantId: restaurant.id,
                menuType: menuType
            )
            self.availableMenus = menus
        } catch {
            errorMessage = "Menüler yüklenirken bir hata oluştu."
        }
        
        isLoading = false
    }
    
    func selectDate(_ date: String) {
        if let index = selectedDates.firstIndex(where: { $0.date == date }) {
            // Deselect
            selectedDates.remove(at: index)
        } else {
            // Check max selections
            if selectedDates.count >= maxDateSelections {
                return
            }
            // Add new date
            selectedDates.append(SelectedDate(date: date, mealTimeSlotId: nil))
        }
    }
    
    func selectMealTimeSlot(for date: String, slotId: Int) {
        if let index = selectedDates.firstIndex(where: { $0.date == date }) {
            selectedDates[index].mealTimeSlotId = slotId
        }
    }
    
    func canProceedToNextStep() -> Bool {
        switch currentStep {
        case 1:
            return selectedDates.count > 0 &&
                   selectedDates.count <= maxDateSelections &&
                   selectedDates.allSatisfy { $0.mealTimeSlotId != nil }
        case 2:
            return selectedRestaurant != nil
        case 3:
            return selectedMenuType != nil
        case 4:
            return selectedMenu != nil
        case 5:
            return true
        default:
            return false
        }
    }
    
    func nextStep() {
        if canProceedToNextStep() && currentStep < 5 {
            if currentStep == 3 {
                // Load menus when moving to step 4
                Task {
                    await loadMenus()
                }
            }
            currentStep += 1
        }
    }
    
    func previousStep() {
        if currentStep > 1 {
            currentStep -= 1
        }
    }
    
    func submitReservation() async {
        guard canProceedToNextStep() else { return }
        
        isLoading = true
        errorMessage = nil
        
        do {
            // Fetch menus for each date
            var menuResults: [[Menu]] = []
            
            for dateObj in selectedDates {
                let menus = try await homeRepository.getMenus(
                    date: dateObj.date,
                    restaurantId: selectedRestaurant!.id,
                    menuType: selectedMenuType!
                )
                menuResults.append(menus)
            }
            
            // Create reservations for each date
            var reservationTasks: [Task<Reservation, Error>] = []
            
            for (index, dateObj) in selectedDates.enumerated() {
                guard let menuForDate = menuResults[index].first else {
                    throw NSError(domain: "ReservationError", code: 1, userInfo: [NSLocalizedDescriptionKey: "Menu not found for date \(dateObj.date)"])
                }
                
                let dateTimeString = "\(dateObj.date)T00:00:00Z"
                
                let request = CreateReservationRequest(
                    restaurantId: selectedRestaurant!.id,
                    menuId: menuForDate.id,
                    mealTimeSlotId: dateObj.mealTimeSlotId!,
                    date: dateTimeString,
                    appetizer: appetizer
                )
                
                let task = Task {
                    try await reservationRepository.createReservation(request)
                }
                reservationTasks.append(task)
            }
            
            // Wait for all reservations to complete
            var reservations: [Reservation] = []
            for task in reservationTasks {
                let reservation = try await task.value
                reservations.append(reservation)
            }
            
            successMessage = "Rezervasyonunuz başarıyla oluşturuldu! \(selectedDates.count) gün için rezervasyon yapıldı."
            
            // Reset form after delay
            DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
                self.resetForm()
            }
            
        } catch {
            errorMessage = "Rezervasyon oluşturulurken bir hata oluştu: \(error.localizedDescription)"
        }
        
        isLoading = false
    }
    
    private func resetForm() {
        selectedDates = []
        selectedRestaurant = nil
        selectedMenuType = nil
        selectedMenu = nil
        selectedMeal = nil
        appetizer = false
        currentStep = 1
    }
}

File: android/feature/make-reservation/src/main/java/com/reservationapp/feature/makereservation/viewmodel/MakeReservationViewModel.kt
package com.reservationapp.feature.makereservation.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.reservationapp.core.common.Result
import com.reservationapp.domain.model.*
import com.reservationapp.domain.repository.HomeRepository
import com.reservationapp.domain.repository.ReservationRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.*

class MakeReservationViewModel(
    private val reservationRepository: ReservationRepository,
    private val homeRepository: HomeRepository
) : ViewModel() {

    private val _currentStep = MutableStateFlow(1)
    val currentStep: StateFlow<Int> = _currentStep.asStateFlow()

    val selectedDates = MutableStateFlow<List<Date>>(emptyList())
    val selectedMealTimeSlot = MutableStateFlow<MealTimeSlot?>(null)

    val selectedRestaurant = MutableStateFlow<Restaurant?>(null)

    val selectedMenuType = MutableStateFlow<MenuType?>(null)

    val selectedMenu = MutableStateFlow<Menu?>(null)
    val selectedMeal = MutableStateFlow<Meal?>(null)
    val appetizer = MutableStateFlow(false)

    private val _mealTimeSlots = MutableStateFlow<List<MealTimeSlot>>(emptyList())
    val mealTimeSlots: StateFlow<List<MealTimeSlot>> = _mealTimeSlots.asStateFlow()

    private val _restaurants = MutableStateFlow<List<Restaurant>>(emptyList())
    val restaurants: StateFlow<List<Restaurant>> = _restaurants.asStateFlow()

    private val _availableMenus = MutableStateFlow<List<Menu>>(emptyList())
    val availableMenus: StateFlow<List<Menu>> = _availableMenus.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _successMessage = MutableStateFlow<String?>(null)
    val successMessage: StateFlow<String?> = _successMessage.asStateFlow()

    init {
        loadInitialData()
    }

    fun loadInitialData() {
        _isLoading.value = true

        viewModelScope.launch {
            val mealTimeSlotsResult = homeRepository.getMealTimeSlots()
            val restaurantsResult = homeRepository.getRestaurants()

            when (mealTimeSlotsResult) {
                is Result.Success -> _mealTimeSlots.value = mealTimeSlotsResult.data
                is Result.Error -> _errorMessage.value = "Öğün saatleri yüklenemedi"
                is Result.Loading -> {}
            }

            when (restaurantsResult) {
                is Result.Success -> _restaurants.value = restaurantsResult.data
                is Result.Error -> _errorMessage.value = "Restoranlar yüklenemedi"
                is Result.Loading -> {}
            }

            _isLoading.value = false
        }
    }

    fun loadMenus() {
        val restaurant = selectedRestaurant.value
        val dates = selectedDates.value
        val menuType = selectedMenuType.value

        if (restaurant == null || dates.isEmpty()) {
            _availableMenus.value = emptyList()
            return
        }

        _isLoading.value = true

        viewModelScope.launch {
            val dateString = formatDateForMenus(dates.first())
            when (val result = homeRepository.getMenus(dateString, restaurant.id)) {
                is Result.Success -> {
                    val filteredMenus = if (menuType != null) {
                        val menuTypeString = when (menuType) {
                            MenuType.STANDARD -> "Standard"
                            MenuType.SPECIAL -> "Special"
                        }
                        result.data.filter { it.menuType == menuTypeString }
                    } else {
                        result.data
                    }
                    _availableMenus.value = filteredMenus
                }
                is Result.Error -> {
                    _errorMessage.value = "Menüler yüklenemedi"
                    _availableMenus.value = emptyList()
                }
                is Result.Loading -> {}
            }
            _isLoading.value = false
        }
    }
    
    private fun formatDateForMenus(date: Date): String {
        val format = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
        return format.format(date)
    }

    fun nextStep() {
        if (canProceedToNextStep()) {
            _currentStep.value = (_currentStep.value + 1).coerceAtMost(5)

            if (_currentStep.value == 4) {
                loadMenus()
            }
        }
    }

    fun previousStep() {
        _currentStep.value = (_currentStep.value - 1).coerceAtLeast(1)
    }

    fun canProceedToNextStep(): Boolean {
        return when (_currentStep.value) {
            1 -> selectedDates.value.isNotEmpty() && selectedMealTimeSlot.value != null
            2 -> selectedRestaurant.value != null
            3 -> selectedMenuType.value != null
            4 -> selectedMenu.value != null
            5 -> true
            else -> false
        }
    }

    fun submitReservation() {
        val restaurant = selectedRestaurant.value
        val menu = selectedMenu.value
        val mealTimeSlot = selectedMealTimeSlot.value
        val dates = selectedDates.value

        if (restaurant == null || menu == null || mealTimeSlot == null || dates.isEmpty()) {
            _errorMessage.value = "Lütfen tüm alanları doldurun"
            return
        }

        _isLoading.value = true
        _errorMessage.value = null

        viewModelScope.launch {
            val dateString = formatDateForApi(dates.first())
            val request = CreateReservationRequest(
                restaurantId = restaurant.id,
                menuId = menu.id,
                mealTimeSlotId = mealTimeSlot.id,
                date = dateString,
                appetizer = appetizer.value
            )

            when (reservationRepository.createReservation(request)) {
                is Result.Success -> {
                    _successMessage.value = "Rezervasyon başarıyla oluşturuldu"
                    resetForm()
                }
                is Result.Error -> {
                    _errorMessage.value = "Rezervasyon oluşturulurken bir hata oluştu."
                }
                is Result.Loading -> {}
            }

            _isLoading.value = false
        }
    }

    private fun resetForm() {
        selectedDates.value = emptyList()
        selectedMealTimeSlot.value = null
        selectedRestaurant.value = null
        selectedMenuType.value = null
        selectedMenu.value = null
        selectedMeal.value = null
        appetizer.value = false
        _currentStep.value = 1
    }

    private fun formatDateForApi(date: Date): String {
        val dateString = formatDateForMenus(date)
        return "${dateString}T00:00:00Z"
    }
}

